packages/context/src/binding-decorator.ts

```diff
@@ -0,0 +1,215 @@
+ * )
+ * export class MyController {
+ * }
+ * ```
+ *
+ * @param templates A list of binding scope/tags or template functions to
+ * configure the binding
+ */
+export function bind(
+  ...templates: (BindingTemplate | BindingScopeAndTags)[]
+): ClassDecorator {
+  const templateFunctions = templates.map(t => {
+    if (typeof t === 'function') {
+      return t;
+    } else {
+      return asBindingTemplate(t);
+    }
+  });
+
+  return (target: Function) => {
+    const spec: BindingMetadata = {
+      templates: [
+        // Add a template to bind to a class
+        binding => {
+          binding.toClass(target as Constructor<unknown>);
+        },
+        ...templateFunctions,
+      ],
+    };
+
+    const decorator = BindDecoratorFactory.createDecorator(
+      BINDING_METADATA_KEY,
+      spec,
+    );
+    decorator(target);
+  };
+}
+
+export namespace bind {
+  /**
+   * `@bind.provider` to denote a provider class
+   *
+   * A list of binding scope/tags or template functions to configure the binding
+   */
+  export function provider(
+    ...templates: (BindingTemplate | BindingScopeAndTags)[]
+  ): ((target: Constructor<Provider<unknown>>) => void) {
+    return (target: Constructor<Provider<unknown>>) => {
+      bind(
+        // Set up the default for providers
+        binding =>
+          binding.toProvider(target).tag('provider', {type: 'provider'}),
+        // Call other template functions
+        ...templates,
+      )(target);
+    };
+  }
+}
+
+/**
+ * Convert binding scope and tags as a template function
+ * @param scopeAndTags
+ */
+function asBindingTemplate(scopeAndTags: BindingScopeAndTags): BindingTemplate {
+  return binding => {
+    if (scopeAndTags.scope) {
+      binding.inScope(scopeAndTags.scope);
+    }
+    if (scopeAndTags.tags) {
+      if (Array.isArray(scopeAndTags.tags)) {
+        binding.tag(...scopeAndTags.tags);
+      } else {
+        binding.tag(scopeAndTags.tags);
+      }
+    }
+  };
+}
+
+/**
+ * Get binding metadata for a class
+ * @param target The target class
+ */
+export function getBindingMetadata(
+  target: Function,
+): BindingMetadata | undefined {
+  return MetadataInspector.getClassMetadata<BindingMetadata>(
+    BINDING_METADATA_KEY,
+    target,
+  );
+}
+
+/**
+ * Get the binding template from a class with binding metadata
+ *
+ * @param cls A class with optional `@bind`
+ */
+export function bindingTemplateFor(cls: Constructor<unknown>): BindingTemplate {
+  const spec = getBindingMetadata(cls);
+  const templateFunctions = (spec && spec.templates) || [];
+  return binding => {
+    for (const t of templateFunctions) {
+      binding.apply(t);
+    }
+    if (spec && spec.target !== cls) {
+      // Remove name/key from base classes
+      binding.apply(removeNameAndKey);
+    }
+  };
+}
+
+export const TYPE_NAMESPACES: {[name: string]: string} = {
+  controller: 'controllers',
+  repository: 'repositories',
```

Comment:
❗️ Context should not need to know about concepts from dependants (repository, controller, etc.), otherwise we have a cyclic dependency, at least at conceptual level: repository depends on context, but context needs to understand repository concepts.

Also, we this hard-coded map is not extensible. How can 3rd-party extensions register new mappings?

Reply:
IMO, Context should provide API allowing other modules to register custom extensions. `@loopback/core`, `@loopback/repository` and `@loopback/service-proxy` should leverage this API to register their type namespaces. The list of namespaces built into `@loopback/context` should contain only `class` and `provider`.

Alternatively, replace `TYPE_NAMESPACES` map with a different mechanism that's easier to extend.